### DISCUSSION TITLE
Does xmake support to re-format compiler output (and how to)?

### DISCUSSION THREAD

**Discussion by Patrick-Ze** [xmake-io/xmake#3663] [Q&A] [answered]
请问xmake是否支持接管编译器的输出（stdout和stderr）并重新格式化后再输出呢？例如，目前在用的一个嵌入式编译器Tasking，会产生类似如下格式的输出
```
ctc W577: ["../source/module.c" 19/34] calling a function without a prototype
0 errors, 1 warnings
```
这个输出格式非常不友好，不被大多数的IDE/编辑器支持，导致无法方便地跳转到代码，并且带有冗余的`0 errors, 1 warnings`信息。我希望达到的效果是，构建工具从编译器获得类似上面这样的输出后，重新进行格式化、移除冗余信息，转换成如下形式，并根据错误等级带上着色：
```
..\source\module.c:19:34: warning: (W577) calling a function without a prototype
```
当前在使用的构建工具是scons，是python编写的，可以通过hook代码来实现此功能，但是启动比较慢，想替换成xmake。请问在xmake中可以实现这样的功能吗？如果可以的话，能否指指路或者给一些提示？
非常感谢~

**Comment by waruqi** [comment, marked as answer]
target 或者搞个 rule ，里面 before_build 加个  os.atexit , 处理下错误。

```
    before_build(function (target)
        os.atexit(function (ok, error)
            -- TODO
        end)
    end)
```

原始错误还是会输出，但通过这个你可以再显示一份格式化的。。

**Reply by Patrick-Ze** [reply]
请问这个`atexit`是只在编译遇到error的时候生效是么？（按照我的理解，报warning一般不会抛出错误码）

**Reply by waruqi** [reply]
程序退出都会执行，通过 ok 判断是否正常退出，error 就是错误信息。。编译 warning 是 xmake -w 实时显示的，不是退出时候显示的

**Reply by Patrick-Ze** [reply]
好的，非常感谢
